src/provider/mod.rs

```diff
@@ -233,3 +233,1513 @@ impl RecommendedFillers for Zksync {
+                value: B256::from_str(
+                    "0x0000000000000000000000000000000000000000000000000000000000000060",
+                )
+                .unwrap(),
+                index: 27900957_u64,
+            }],
+        });
+        run_server_and_test(
+            |module| {
+                let proof_rpc_response = proof.clone();
+                let keys_rpc_request = keys.clone();
+                module
+                    .register_method::<RpcResult<Option<Proof>>, _>(
+                        "zks_getProof",
+                        move |params, _, _| {
+                            let (address_param, keys_param, batch_num_param) =
+                                params.parse::<(Address, Vec<B256>, u64)>().unwrap();
+                            assert_eq!(address_param, address);
+                            assert_eq!(keys_param, keys_rpc_request);
+                            assert_eq!(batch_num_param, l1_batch_number);
+                            Ok(proof_rpc_response.clone())
+                        },
+                    )
+                    .unwrap();
+            },
+            |http_url: reqwest::Url| async {
+                let provider = zksync_provider()
+                    .with_recommended_fillers()
+                    .on_http(http_url);
+
+                let received_proof =
+                    ZksyncProvider::get_proof(&provider, address, keys.clone(), l1_batch_number)
+                        .await
+                        .unwrap();
+                assert_eq!(proof, received_proof);
+            },
+        )
+        .await;
+    }
+
+    #[tokio::test(flavor = "multi_thread")]
+    async fn get_proof_when_not_available() {
+        let address = address!("0000000000000000000000000000000000008003");
+        let keys = vec![B256::from_str(
+            "0x8b65c0cf1012ea9f393197eb24619fd814379b298b238285649e14f936a5eb12",
+        )
+        .unwrap()];
+        let l1_batch_number = 354895_u64;
+        run_server_and_test(
+            |module| {
+                let keys_rpc_request = keys.clone();
+                module
+                    .register_method::<RpcResult<Option<Proof>>, _>(
+                        "zks_getProof",
+                        move |params, _, _| {
+                            let (address_param, keys_param, batch_num_param) =
+                                params.parse::<(Address, Vec<B256>, u64)>().unwrap();
+                            assert_eq!(address_param, address);
+                            assert_eq!(keys_param, keys_rpc_request);
+                            assert_eq!(batch_num_param, l1_batch_number);
+                            Ok(None)
+                        },
+                    )
+                    .unwrap();
+            },
+            |http_url: reqwest::Url| async {
+                let provider = zksync_provider()
+                    .with_recommended_fillers()
+                    .on_http(http_url);
```

Comment:
AFAICS we always instantiate the provider this way, so it might make sense to change `run_server_and_test` so that it instantiates the provider for you and gives as a function argument to the closure instead of URL.

Reply:
Good idea, I did as you suggested.